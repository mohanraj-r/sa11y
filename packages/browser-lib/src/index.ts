/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { base, recommended, full } from '@sa11y/preset-rules';
export { exceptionListFilter } from '@sa11y/format';
export { assertAccessible } from '@sa11y/assert';

export const nameSpace = 'sa11y';
