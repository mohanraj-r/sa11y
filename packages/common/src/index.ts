/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getViolations } from './axe';
export { AssertFunction, Options, WDIOBrowser } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
