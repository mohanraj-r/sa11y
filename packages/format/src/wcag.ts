/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

type WcagLevel = 'A' | 'AA' | 'AAA' | undefined;
type WcagVersion = '2.0' | '2.1' | undefined;

/**
 * Process given tags from a11y violations and return WCAG meta-data
 * Ref: https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#axe-core-tags
 */
export class WcagMetadata {
    static readonly regExp = /^(wcag)(?<version_or_sc>\d+)(?<level>a*)$/;
    static readonly defaultSC = 'best-practice';
    public wcagLevel: WcagLevel;
    public wcagVersion: WcagVersion;
    public successCriteria = WcagMetadata.defaultSC; // Default SC for non-wcag rules
    constructor(readonly tags: string[]) {
        tags.forEach((tag) => {
            const match = WcagMetadata.regExp.exec(tag);
            if (match && match.groups) {
                const level = match.groups.level;
                // Tags starting with "wcag" can contain either wcag version and level
                // or success criteria
                if (level) {
                    this.wcagLevel = level.toUpperCase() as WcagLevel;
                    if (match.groups.version_or_sc === '2') {
                        this.wcagVersion = '2.0'; // Add decimal for consistency
                    } else {
                        this.wcagVersion = match.groups.version_or_sc.split('').join('.') as WcagVersion;
                    }
                } else {
                    this.successCriteria = match.groups.version_or_sc.split('').join('.');
                }
            }
        });
    }

    /**
     * Return formatted string containing WCAG version, level and SC
     */
    toString(): string {
        if (!this.wcagVersion || !this.wcagLevel) {
            throw new Error(`Unable to set WCAG version and level from given tags: ${this.tags.join(', ')}`);
        }
        return `WCAG-${this.wcagVersion}-Level-${this.wcagLevel} SC-${this.successCriteria}`;
    }
}
