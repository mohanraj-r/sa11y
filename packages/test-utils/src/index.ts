/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    domWithA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    a11yIssuesCount,
    shadowDomID,
    videoURL,
} from './test-data';
export { beforeEachSetup, cartesianProduct, checkA11yError } from './utils';
