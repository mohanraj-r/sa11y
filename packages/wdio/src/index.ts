/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { assertAccessible, assertAccessibleSync, Options } from './wdio';
export { setup } from './setup';
