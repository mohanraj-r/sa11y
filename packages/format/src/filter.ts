/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults, ExceptionList } from '@sa11y/common';

/**
 * Filter a11y violations from axe based on given {@link ExceptionList}
 * @param violations - List of violations found with axe
 * @param exceptionList - {@link ExceptionList} of map of rule to corresponding css targets that needs to be filtered from a11y results
 */
export function exceptionListFilter(violations: AxeResults, exceptionList: ExceptionList = {}): AxeResults {
    const exceptionRules = Object.keys(exceptionList);
    if (exceptionRules.length === 0) return violations;

    const filteredViolations: AxeResults = [];

    for (const violation of violations) {
        if (!exceptionRules.includes(violation.id)) {
            filteredViolations.push(violation);
        } else {
            for (const result of violation.nodes) {
                const filteredResults = result.target.filter(
                    (cssSelector) => !exceptionList[violation.id].includes(cssSelector)
                );
                if (filteredResults.length > 0) {
                    filteredViolations.push(violation);
                }
            }
        }
    }

    return filteredViolations;
}
