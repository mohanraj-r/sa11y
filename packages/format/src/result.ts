/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from '@sa11y/common';
import { NodeResult, Result } from 'axe-core';
import * as axe from 'axe-core';
import { WcagMetadata } from './wcag';

const defaultImpact = 'minor'; // if impact is undefined
// Helper object to sort violations by impact order
const impactOrder = {
    critical: 1,
    serious: 2,
    moderate: 3,
    minor: 4,
};

/**
 * Filtered a11y result containing essential info about the a11y failure
 */
export class A11yResult {
    public readonly id: string;
    public readonly selectors: string;
    public readonly description: string;
    public readonly helpUrl: string;
    // TODO: Add WCAG levels to sort rule
    public readonly wcag: WcagMetadata;
    // public readonly testPath: string;
    // public readonly testNames: string[];

    /**
     * Normalize and flatten a11y violations from Axe
     */
    static convert(violations: AxeResults): A11yResult[] {
        return A11yResult.sort(violations).flatMap((violation) => {
            return violation.nodes.map((node) => {
                return new A11yResult(violation, node);
            });
        });
    }

    /**
     * Sorts give a11y violations from axe in order of impact
     */
    static sort(violations: AxeResults): AxeResults {
        return violations.sort((a, b) => {
            const aImpact = impactOrder[a.impact || defaultImpact];
            const bImpact = impactOrder[b.impact || defaultImpact];
            if (aImpact < bImpact) return -1;
            if (aImpact > bImpact) return 1;
            return 0;
        });
    }

    constructor(violation: Result, node: NodeResult) {
        this.id = violation.id;
        this.description = violation.help;
        this.wcag = new WcagMetadata(violation.tags);
        this.helpUrl = violation.helpUrl.split('?')[0];
        this.selectors = node.target.sort().join('; ');
    }
}

/**
 * Consolidate unique a11y violations by removing duplicates.
 */
export class ConsolidatedResults {
    static a11yResults: Record<string, A11yResult[]> = {};
    static consolidatedMap = new Map<string, axe.Result>();

    static clear(): void {
        this.consolidatedMap.clear();
    }

    /**
     * Convert and consolidate given a11y results based on given key
     */
    // TODO(refactor): Merge with ConsolidatedResults.add()
    static convert(results: AxeResults, key = ''): void {
        // const a11yResults = ConsolidatedResults.a11yResults;
        // // Initialize if key doesn't exist
        // if (!Array.isArray(a11yResults[key])) a11yResults[key] = [];
        // for (const result of results) {
        //     for (const node of result.nodes) {
        //         a11yResults[key].push(<A11yResult>{
        //             // TODO (refactor): Reuse A11yResult for A11yError.format()
        //             id: result.id,
        //             // Note: Use a separator that cannot be part of a CSS selector
        //             selectors: node.target.sort().join('; '),
        //             description: result.help,
        //             helpUrl: result.helpUrl.split('?')[0],
        //             impact: result.impact,
        //             // TODO : Add wcag level or best practice tag to output ?
        //             // const criteria = violation.tags.filter((tag) => tag.startsWith('wcag2a') || tag.startsWith('best'));
        //         });
        //     }
        // }
    }

    /**
     * Adds given a11y results to a consolidated list if they are not already present
     * @returns results that have not been added earlier
     */
    static add(results: AxeResults): AxeResults {
        const newResults: AxeResults = [];
        for (const result of results) {
            for (const node of result.nodes) {
                const selectors = node.target.sort().join(';');
                const key = `${result.id}--${selectors}`;
                if (!this.consolidatedMap.has(key)) {
                    this.consolidatedMap.set(key, result);
                    // TODO (refactor): Store A11yResult with selected fields instead
                    //  AND reuse it to display formatted error message
                    newResults.push(result);
                }
            }
        }
        return newResults;
    }
}
