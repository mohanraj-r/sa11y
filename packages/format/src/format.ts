/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { AxeResults, errMsgHeader } from '@sa11y/common';
import { ConsolidatedResults } from './result';

/**
 * Custom formatter to format a11y violations found by axe
 * Use `JSON.stringify` to return violations without formatting
 */
export interface Formatter {
    (violations: AxeResults): string;
}

/**
 * Highlights the import part of the error message
 */
export interface Highlighter {
    (text: string): string;
}

/**
 * Optional parameters used while formatting a11y issues
 */
export interface Options {
    a11yViolationIndicator: string;
    helpUrlIndicator: string;
    formatter?: Formatter;
    highlighter: Highlighter;
}

/**
 * Default options to be used while formatting a11y issues
 */
const defaultOptions: Options = {
    a11yViolationIndicator: '*',
    helpUrlIndicator: '-',
    // TODO (refactor): Create a Default formatter that points to A11yError.format()
    formatter: undefined,
    highlighter: (text: string): string => text,
};

const defaultImpact = 'minor'; // if impact is undefined
// Helper object to sort violations by impact order
const impactOrder = {
    critical: 1,
    serious: 2,
    moderate: 3,
    minor: 4,
};

/**
 * Sorts give a11y results from axe in order of impact
 */
export function sortViolations(violations: AxeResults): void {
    violations.sort((a, b) => {
        const aImpact = impactOrder[a.impact || defaultImpact];
        const bImpact = impactOrder[b.impact || defaultImpact];
        if (aImpact < bImpact) return -1;
        if (aImpact > bImpact) return 1;
        return 0;
    });
}

/**
 *  Custom error object to represent a11y violations
 */
export class A11yError extends Error {
    /**
     * Throw error with formatted a11y violations
     * @param violations - List of a11y violations
     * @param consolidate - Filter our previously reported issues and report only new
     *  issues which haven't been previously reported.
     * @param opts - Options used for formatting a11y issues
     */
    static checkAndThrow(violations: AxeResults, consolidate = false, opts: Partial<Options> = defaultOptions): void {
        if (consolidate) {
            violations = ConsolidatedResults.add(violations);
            // TODO (debug): Will this affect all errors globally?
            Error.stackTraceLimit = 0;
        }
        if (violations.length > 0) {
            throw new A11yError(violations, opts);
        }
    }

    constructor(readonly violations: AxeResults, opts: Partial<Options> = defaultOptions) {
        super(`${violations.length} ${errMsgHeader}`);
        this.name = A11yError.name;
        this.message = `${violations.length} ${errMsgHeader}\n ${this.format(opts)}`;
    }

    get length(): number {
        return this.violations.length;
    }

    /**
     * Format a11y violations into a readable format highlighting important information to help fixing the issue.
     * @param opts - Options used for formatting a11y issues.
     */
    format(opts: Partial<Options> = defaultOptions): string {
        const options = Object.assign(Object.assign({}, defaultOptions), opts);
        if (options.formatter !== undefined) {
            return options.formatter(this.violations);
        }

        sortViolations(this.violations);
        return this.violations
            .map((violation) => {
                return violation.nodes
                    .map((node) => {
                        // Note: Use a separator that cannot be part of a CSS selector
                        const selectors = node.target.join('; ');
                        const helpURL = violation.helpUrl.split('?')[0];
                        // TODO : Add wcag level or best practice tag to output ?
                        // const criteria = violation.tags.filter((tag) => tag.startsWith('wcag2a') || tag.startsWith('best'));

                        return (
                            options.highlighter(
                                `${options.a11yViolationIndicator} (${violation.id}) ${violation.help}: ${selectors}`
                            ) + `\n\t${options.helpUrlIndicator} Help URL: ${helpURL}`
                        );
                    })
                    .join('\n\n');
            })
            .join('\n\n');
    }
}
