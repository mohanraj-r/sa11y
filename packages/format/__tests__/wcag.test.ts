/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { WcagMetadata } from '../src/';

// input tags, expected version, expected level, expected SC
const noErrorCases = [
    [['wcag2a'], '2.0', 'A', WcagMetadata.defaultSC],
    [['wcag21aaa'], '2.1', 'AAA', WcagMetadata.defaultSC],
    [['wcag222', 'wcag21aa'], '2.1', 'AA', '2.2.2'],
];

const errorCases = [
    [[], undefined, undefined, WcagMetadata.defaultSC],
    [[WcagMetadata.defaultSC], undefined, undefined, WcagMetadata.defaultSC],
    [['wcag222'], undefined, undefined, '2.2.2'],
    [['foo', 'bar'], undefined, undefined, WcagMetadata.defaultSC],
    [['wcag2foo', 'wcag21bar'], undefined, undefined, WcagMetadata.defaultSC],
];

describe('WCAG Metadata extractor', () => {
    it.each([...noErrorCases, ...errorCases])(
        'should extract WCAG version and level (arg:%p)',
        (tags: string[], wcagVersion: string, wcagLevel: string, successCriteria: string) => {
            const wcag = new WcagMetadata(tags);
            expect(wcag.wcagVersion).toBe(wcagVersion);
            expect(wcag.wcagLevel).toBe(wcagLevel);
            expect(wcag.successCriteria).toBe(successCriteria);
        }
    );

    it.each(noErrorCases)(
        'should format WCAG metadata when required data is present (arg:%p)',
        (tags: string[], wcagVersion: string, wcagLevel: string, successCriteria: string) => {
            const wcag = new WcagMetadata(tags).toString();
            expect(wcag).toMatchSnapshot();
            expect(wcag).toContain(wcagVersion);
            expect(wcag).toContain(wcagLevel);
            expect(wcag).toContain(successCriteria);
        }
    );
});
