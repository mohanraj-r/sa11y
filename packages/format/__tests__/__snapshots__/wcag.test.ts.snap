// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag2a"]) 1`] = `"WCAG-2.0-Level-A SC-best-practice"`;

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag21aaa"]) 1`] = `"WCAG-2.1-Level-AAA SC-best-practice"`;

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag222", "wcag21aa"]) 1`] = `"WCAG-2.1-Level-AA SC-2.2.2"`;

exports[`WCAG Metadata extractor should throw error when required data is not present (arg:["foo", "bar"]) 1`] = `"Unable to set WCAG version and level from given tags: foo, bar"`;

exports[`WCAG Metadata extractor should throw error when required data is not present (arg:["wcag2foo", "wcag21bar"]) 1`] = `"Unable to set WCAG version and level from given tags: wcag2foo, wcag21bar"`;

exports[`WCAG Metadata extractor should throw error when required data is not present (arg:["wcag222"]) 1`] = `"Unable to set WCAG version and level from given tags: wcag222"`;

exports[`WCAG Metadata extractor should throw error when required data is not present (arg:[]) 1`] = `"Unable to set WCAG version and level from given tags: "`;
